src/proxy/inbound.rs

```diff
@@ -365,26 +460,65 @@ impl Inbound {
             upstream_addr,
             inbound_protocol,
             hbone_addr,
+            upstream_protocol_addr,
         })
     }
 
+    // Selects a service by hostname without the explicit knowledge of the namespace
+    // There is no explicit mapping from hostname to namespace (e.g. foo.com)
+    async fn find_service_by_hostname(
+        state: &DemandProxyState,
+        local_workload: &Workload,
+        hbone_host: Strng,
+    ) -> Result<Service, Error> {
+        // Validate a service exists for the hostname
+        // TODO(jaellio): Currently returns an error is no service is found. Workload lookup is not supported.
+        let services = state.read().find_service_by_hostname(&hbone_host)?;
+
+        services
+            .iter()
+            .max_by_key(|s| {
+                let is_local_namespace = s.namespace == local_workload.namespace;
+                match is_local_namespace {
+                    true => 1,
+                    false => 0,
+                }
+            })
+            .cloned()
+            .ok_or_else(|| Error::NoResolvedAddresses(hbone_host.to_string()))
```

Comment:
`no ip addresses were resolved for workload: {0}` is the error message, but this is a service name which seems a bit awkward. May be worth a new enum for this case?

Reply:
Updated